src/category_theory/connected.lean

```diff
@@ -0,0 +1,375 @@
+
+/--
+An inductive-like property for the objects of a connected category.
+If `default J` is in the set `p`, and `p` is closed under morphisms of `J`,
+then `p` contains all of `J`.
+
+The converse is given in `connected.of_induct`.
+-/
+lemma induct_on_objects [connected J] (p : set J) (h0 : default J ∈ p)
+  (h1 : ∀ {j₁ j₂ : J} (f : j₁ ⟶ j₂), j₁ ∈ p ↔ j₂ ∈ p) (j : J) :
+  j ∈ p :=
+begin
+  injection (constant_function_of_preserves_morphisms (λ k, ulift.up (k ∈ p)) (λ j₁ j₂ f, _) j) with i,
+  rwa i,
+  dsimp,
+  exact congr_arg ulift.up (propext (h1 f)),
+end
+
+/--
+If any maximal connected component of J containing the default is all of J, then J is connected.
+
+The converse of `induct_on_objects`.
+-/
+def connected.of_induct [inhabited J]
+  (h : ∀ (p : set J), default J ∈ p → (∀ {j₁ j₂ : J} (f : j₁ ⟶ j₂), j₁ ∈ p ↔ j₂ ∈ p) → ∀ (j : J), j ∈ p) :
+  connected J :=
+connected.of_constant_of_preserves_morphisms (λ α F a, h {j | F j = F (default J)} rfl (λ _ _ f, by simp [a f] ))
+
+/-- j₁ and j₂ are related by `zag` if there is a morphism between them. -/
+@[reducible]
+def zag (j₁ j₂ : J) : Prop := nonempty (j₁ ⟶ j₂) ∨ nonempty (j₂ ⟶ j₁)
+/--
+`j₁` and `j₂` are related by `zigzag` if there is a chain of
+morphisms from `j₁` to `j₂`, with backward morphisms allowed.
+-/
+@[reducible]
+def zigzag : J → J → Prop := relation.refl_trans_gen zag
+
+/-- Any equivalence relation containing (⟶) holds for all pairs of a connected category. -/
+lemma equiv_relation [connected J] (r : J → J → Prop) (hr : _root_.equivalence r)
+  (h : ∀ {j₁ j₂ : J} (f : j₁ ⟶ j₂), r j₁ j₂) :
+  ∀ (j₁ j₂ : J), r j₁ j₂ :=
+begin
+  have z: ∀ (j : J), r (default J) j :=
+    induct_on_objects (λ k, r (default J) k)
+        (hr.1 (default J)) (λ _ _ f, ⟨λ t, hr.2.2 t (h f), λ t, hr.2.2 t (hr.2.1 (h f))⟩),
+  intros, apply hr.2.2 (hr.2.1 (z _)) (z _)
+end
+
+/-- In a connected category, any two objects are related by `zigzag`. -/
+lemma connected_zigzag [connected J] (j₁ j₂ : J) : zigzag j₁ j₂ :=
+equiv_relation _
+  (mk_equivalence _
+    relation.reflexive_refl_trans_gen
+    (relation.refl_trans_gen.symmetric (λ _ _ _, by rwa [zag, or_comm]))
+    relation.transitive_refl_trans_gen)
+  (λ _ _ f, relation.refl_trans_gen.single (or.inl (nonempty.intro f))) _ _
+
+/--
+If any two objects in an inhabited category are related by `zigzag`, the category is connected.
+-/
+def zigzag_connected [inhabited J] (h : ∀ (j₁ j₂ : J), zigzag j₁ j₂) : connected J :=
+begin
+  apply connected.of_induct,
+  intros,
+  have: ∀ (j₁ j₂ : J), zigzag j₁ j₂ → (j₁ ∈ p ↔ j₂ ∈ p),
+    introv k,
+    apply relation.refl_trans_gen.head_induction_on k,
+    refl,
+    intros,
+    rw ← a_3,
+    rcases h' with ⟨⟨_⟩⟩ | ⟨⟨_⟩⟩,
+    apply a_1 h',
+    symmetry,
+    apply a_1 h',
+  rwa this j (default J) (h _ _),
+end
+
+omit 𝒥
+
+/-- Analogous to `last`. -/
+def head' {α : Type v₂} : Π l : list α, l ≠ list.nil → α
+| [] t := absurd rfl t
+| (a :: l) _ := a
```

Comment:
Move to `data/list/defs.lean`. Needs a better doc-string. Perhaps rename to `head_of_ne_nil`?